flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java

```diff
@@ -1255,4 +1268,17 @@ <T extends State> T transitionToState(StateFactory<T> targetState) {
     State getState() {
         return state;
     }
+
+    /**
+     * Check for slots that are idle for more than {@link JobManagerOptions#SLOT_IDLE_TIMEOUT} and
+     * release them back to the ResourceManager.
+     */
+    private void checkIdleSlotTimeout() {
+        declarativeSlotPool.releaseIdleSlots(System.currentTimeMillis());
+        getMainThreadExecutor()
+                .schedule(
+                        this::checkIdleSlotTimeout,
+                        slotIdleTimeout.toMillis(),
+                        TimeUnit.MILLISECONDS);
+    }
```

Comment:
Consider only running this in particular states; as-is this will run infinitely, even after the scheduler was shut down.

Reply:
I was always under the impression that the main thread executor would get shut down with the component 🤔 Can you please confirm this? This will indeed leak if that's not the case.

Reply:
The main thread executor is scoped to a JobMaster. While it won't run infinitely it will still run while the scheduler was shut down but the JM is still running (e.g., during the JM shutdown procedure)